File: scripts/check_balance.py
"""
DeepSeek 余额查询 + QQ 通知工具
通过 GitHub Action 每日定时执行，查询 DeepSeek API 余额并推送到 QQ。

需要设置以下 GitHub Secrets / 环境变量：
  - DEEPSEEK_API_KEY: DeepSeek API 密钥
  - QMSG_KEY: Qmsg 酱的 KEY（从 https://qmsg.zendee.cn 获取）
  - QMSG_QQ: 接收消息的 QQ 号（可选，默认为 Qmsg 酱后台配置的默认号码）
"""

import os
import sys
import json
import requests
from datetime import datetime, timezone, timedelta

# ============================================================
# 配置
# ============================================================

DEEPSEEK_API_KEY = os.environ.get("DEEPSEEK_API_KEY", "")
QMSG_KEY = os.environ.get("QMSG_KEY", "")
QMSG_QQ = os.environ.get("QMSG_QQ", "")  # 可选，指定接收 QQ 号

# DeepSeek API 端点
DEEPSEEK_BALANCE_URL = "https://api.deepseek.com/user/balance"

# Qmsg 酱 API 端点
QMSG_SEND_URL = "https://qmsg.zendee.cn/api/v2/send"

# 北京时区
TZ_BEIJING = timezone(timedelta(hours=8))

# ============================================================
# 颜色/Emoji 辅助
# ============================================================

def status_emoji(percent: float) -> str:
    """根据余额百分比返回对应的 emoji"""
    if percent > 50:
        return "🟢"
    elif percent > 20:
        return "🟡"
    elif percent > 5:
        return "🟠"
    else:
        return "🔴"


# ============================================================
# DeepSeek 余额查询
# ============================================================

def fetch_deepseek_balance() -> dict:
    """查询 DeepSeek 账户余额，返回结构化数据"""
    if not DEEPSEEK_API_KEY:
        return {"error": "未设置 DEEPSEEK_API_KEY 环境变量"}

    headers = {
        "Authorization": f"Bearer {DEEPSEEK_API_KEY}",
        "Content-Type": "application/json",
    }

    try:
        resp = requests.get(DEEPSEEK_BALANCE_URL, headers=headers, timeout=15)
        resp.raise_for_status()
        data = resp.json()

        # DeepSeek 余额接口返回格式:
        # {
        #   "is_available": true,
        #   "balance_infos": [
        #     {"currency": "CNY", "total_balance": "100.00", "topped_up_balance": "100.00", "granted_balance": "0.00"}
        #   ]
        # }
        return {"success": True, "data": data}

    except requests.exceptions.Timeout:
        return {"error": "请求 DeepSeek API 超时"}
    except requests.exceptions.HTTPError as e:
        return {"error": f"HTTP 错误: {e.response.status_code}", "detail": str(e)}
    except requests.exceptions.RequestException as e:
        return {"error": f"网络请求失败: {str(e)}"}
    except json.JSONDecodeError:
        return {"error": "API 返回数据解析失败"}


# ============================================================
# 格式化消息
# ============================================================

def format_balance_message(result: dict) -> str:
    """将查询结果格式化为可读的消息"""
    now = datetime.now(TZ_BEIJING).strftime("%Y-%m-%d %H:%M:%S")

    if "error" in result:
        return (
            f"⚠️ DeepSeek 余额查询失败\n"
            f"时间: {now}\n"
            f"原因: {result['error']}"
        )

    data = result.get("data", {})

    lines = [
        f"💰 DeepSeek 余额日报",
        f"📅 {now}",
        f"{'─' * 20}",
    ]

    balance_infos = data.get("balance_infos", [])
    if not balance_infos:
        lines.append("⚠️ 未获取到余额信息")
        lines.append(f"原始返回: {json.dumps(data, ensure_ascii=False)}")
        return "\n".join(lines)

    for info in balance_infos:
        currency = info.get("currency", "未知")
        total = float(info.get("total_balance", 0))
        topped_up = float(info.get("topped_up_balance", 0))
        granted = float(info.get("granted_balance", 0))
        used = topped_up + granted - total  # 推算已使用额度

        lines.append(f"币种: {currency}")
        lines.append(f"总余额: {total:.4f}")
        lines.append(f"  充值余额: {topped_up:.4f}")
        lines.append(f"  赠送余额: {granted:.4f}")

        if topped_up > 0:
            remaining_percent = (total / topped_up * 100) if topped_up > 0 else 0
            emoji = status_emoji(remaining_percent)
            lines.append(f"  已使用: {used:.4f}")
            lines.append(f"  剩余比例: {emoji} {remaining_percent:.1f}%")

    is_available = data.get("is_available", True)
    if not is_available:
        lines.append("\n🚫 账户余额不足，API 不可用！")

    return "\n".join(lines)


# ============================================================
# QQ 通知（通过 Qmsg 酱）
# ============================================================

def send_qq_message(message: str) -> dict:
    """通过 Qmsg 酱发送消息到 QQ"""
    if not QMSG_KEY:
        return {"error": "未设置 QMSG_KEY 环境变量"}

    payload = {
        "msg": message,
    }
    if QMSG_QQ:
        payload["qq"] = QMSG_QQ

    try:
        resp = requests.post(
            f"{QMSG_SEND_URL}/{QMSG_KEY}",
            data=payload,
            timeout=15,
        )
        resp.raise_for_status()
        result = resp.json()

        if result.get("success"):
            return {"success": True, "data": result}
        else:
            return {"error": result.get("reason", "Qmsg 返回失败")}

    except requests.exceptions.Timeout:
        return {"error": "发送 QQ 消息超时"}
    except requests.exceptions.RequestException as e:
        return {"error": f"发送 QQ 消息失败: {str(e)}"}


# ============================================================
# 备选通知：Server 酱（如果 Qmsg 失败，可通过 Server 酱发微信）
# ============================================================

def send_serverchan_message(title: str, message: str) -> dict:
    """通过 Server 酱发送消息到微信（备选方案）"""
    sckey = os.environ.get("SERVERCHAN_KEY", "")
    if not sckey:
        return {"error": "未设置 SERVERCHAN_KEY"}

    try:
        resp = requests.post(
            f"https://sctapi.ftqq.com/{sckey}.send",
            data={"title": title, "desp": message},
            timeout=15,
        )
        resp.raise_for_status()
        return {"success": True, "data": resp.json()}
    except Exception as e:
        return {"error": str(e)}


# ============================================================
# 主流程
# ============================================================

def main():
    print("=" * 50)
    print("DeepSeek 余额查询工具")
    print("=" * 50)

    # 1. 查询余额
    print("\n[1/3] 正在查询 DeepSeek 余额...")
    balance_result = fetch_deepseek_balance()

    if "error" in balance_result:
        print(f"  ❌ {balance_result['error']}")
    else:
        print(f"  ✅ 查询成功")

    # 2. 格式化消息
    print("\n[2/3] 正在格式化消息...")
    message = format_balance_message(balance_result)
    print(message)

    # 3. 发送 QQ 通知
    print("\n[3/3] 正在发送 QQ 通知...")
    qq_result = send_qq_message(message)

    if "error" in qq_result:
        print(f"  ❌ QQ 通知失败: {qq_result['error']}")

        # 备选：Server 酱
        sc_result = send_serverchan_message("DeepSeek 余额", message)
        if "error" in sc_result:
            print(f"  ❌ 备选通知也失败: {sc_result['error']}")
        else:
            print(f"  ℹ️  已通过 Server 酱发送到微信")
    else:
        print(f"  ✅ QQ 通知发送成功")

    # 最终状态
    if "error" in balance_result:
        print("\n⚠️  任务完成（余额查询失败）")
        sys.exit(1)
    else:
        print("\n✅ 任务完成")
        sys.exit(0)


if __name__ == "__main__":
    main()

File: scripts/check_balance.mjs
/**
 * DeepSeek 余额查询 + 手机通知工具（Node.js 零依赖版）
 * 通过 GitHub Action 每日定时执行，查询 DeepSeek API 余额并通过 Server 酱3 推送通知。
 *
 * 需要设置以下 GitHub Secrets / 环境变量：
 *   - DEEPSEEK_API_KEY: DeepSeek API 密钥
 *   - SERVER_UID: Server 酱3 用户 UID
 *   - SERVER_KEY: Server 酱3 SendKey
 */

// ============================================================
// 配置
// ============================================================

const DEEPSEEK_API_KEY = process.env.DEEPSEEK_API_KEY || "";
const SERVER_UID = process.env.SERVER_UID || "";
const SERVER_KEY = process.env.SERVER_KEY || "";

const DEEPSEEK_BALANCE_URL = "https://api.deepseek.com/user/balance";

// ============================================================
// 工具函数
// ============================================================

function nowBeijing() {
  return new Date().toLocaleString("zh-CN", { timeZone: "Asia/Shanghai" });
}

function statusEmoji(percent) {
  if (percent > 50) return "🟢";
  if (percent > 20) return "🟡";
  if (percent > 5) return "🟠";
  return "🔴";
}

// ============================================================
// DeepSeek 余额查询
// ============================================================

async function fetchDeepSeekBalance() {
  if (!DEEPSEEK_API_KEY) {
    return { error: "未设置 DEEPSEEK_API_KEY 环境变量" };
  }

  try {
    const resp = await fetch(DEEPSEEK_BALANCE_URL, {
      headers: {
        Authorization: `Bearer ${DEEPSEEK_API_KEY}`,
        "Content-Type": "application/json",
      },
      signal: AbortSignal.timeout(15000),
    });

    const rawText = await resp.text();
    console.log(`[DEBUG] DeepSeek API 原始返回: ${rawText}`);

    if (!resp.ok) {
      return { error: `HTTP ${resp.status}`, detail: rawText };
    }

    const data = JSON.parse(rawText);
    return { success: true, data, raw: rawText };
  } catch (e) {
    if (e.name === "TimeoutError" || e.name === "AbortError") {
      return { error: "请求 DeepSeek API 超时" };
    }
    return { error: `网络请求失败: ${e.message}` };
  }
}

// ============================================================
// 格式化消息
// ============================================================

function formatBalanceMessage(result) {
  const now = nowBeijing();

  if (result.error) {
    return `⚠️ DeepSeek 余额查询失败\n时间: ${now}\n原因: ${result.error}`;
  }

  const data = result.data || {};
  const balanceInfos = data.balance_infos || [];

  const lines = [`📅 ${now}`, `\nAPI 原始返回:\n${result.raw || JSON.stringify(data)}`];

  if (!balanceInfos.length) {
    lines.push("⚠️ 未获取到余额信息");
    return lines.join("\n");
  }

  for (const info of balanceInfos) {
    const currency = info.currency || "未知";
    const total = parseFloat(info.total_balance || 0);
    const toppedUp = parseFloat(info.topped_up_balance || 0);
    const granted = parseFloat(info.granted_balance || 0);
    const used = toppedUp + granted - total;

    lines.push(`\n💰 ${currency}`);
    lines.push(`总余额: ${total.toFixed(4)}`);
    lines.push(`充值余额: ${toppedUp.toFixed(4)}`);
    lines.push(`赠送余额: ${granted.toFixed(4)}`);
    lines.push(`已使用: ${used.toFixed(4)}`);
  }

  if (data.is_available === false) {
    lines.push("\n🚫 账户余额不足，API 不可用！");
  }

  return lines.join("\n");
}

// ============================================================
// Server 酱3 通知
// ============================================================

async function sendServerChanMessage(title, desp) {
  if (!SERVER_UID || !SERVER_KEY) {
    return { error: "未设置 SERVER_UID 或 SERVER_KEY" };
  }

  try {
    const url = `https://${SERVER_UID}.push.ft07.com/send/${SERVER_KEY}.send`;
    const resp = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: new URLSearchParams({ title, desp }),
      signal: AbortSignal.timeout(15000),
    });

    const result = await resp.json();
    console.log(`[DEBUG] Server 酱返回: ${JSON.stringify(result)}`);

    if (result.code === 0) {
      return { success: true, data: result };
    }
    return { error: result.message || "Server酱3 返回失败" };
  } catch (e) {
    if (e.name === "TimeoutError" || e.name === "AbortError") {
      return { error: "Server酱3 请求超时" };
    }
    return { error: `Server酱3 请求失败: ${e.message}` };
  }
}

// ============================================================
// 主流程
// ============================================================

async function main() {
  console.log("=".repeat(50));
  console.log("DeepSeek 余额查询工具 (Node.js)");
  console.log("=".repeat(50));

  // 1. 查询余额
  console.log("\n[1/3] 正在查询 DeepSeek 余额...");
  const balanceResult = await fetchDeepSeekBalance();

  if (balanceResult.error) {
    console.log(`  ❌ ${balanceResult.error}`);
  } else {
    console.log("  ✅ 查询成功");
  }

  // 2. 格式化消息
  console.log("\n[2/3] 正在格式化消息...");
  const message = formatBalanceMessage(balanceResult);
  console.log(message);

  // 3. 发送通知
  console.log("\n[3/3] 正在发送通知...");
  console.log("  → Server 酱...");
  const scResult = await sendServerChanMessage("DeepSeek 余额", message);
  if (!scResult.error) {
    console.log("  ✅ Server 酱发送成功");
  } else {
    console.log(`  ❌ Server 酱失败: ${scResult.error}`);
  }

  // 最终状态
  if (balanceResult.error) {
    console.log("\n⚠️  任务完成（余额查询失败）");
    process.exit(1);
  } else {
    console.log("\n✅ 任务完成");
    process.exit(0);
  }
}

main();
